rep-0003.rst

```diff
@@ -281,6 +281,68 @@ Build System Support:
 
 - Same as Indigo
 
+Melodic Morenia (May 2018 - May 2023)
+--------------------------------------
+Required Support for:
+
+- Ubuntu Artful (17.10)
+- Ubuntu B (18.04)
+
+Recommended Support for:
+
+- Debian Stretch
+- Fedora 28
+
+Architectures Supported:
+
+- amd64
+- arm32
+- arm64
+
+Targeted Languages:
+
+- C++11
```

Comment:
Should we now change the motivation and/or recommendation about using C++11 in desktop-full packages? Since we have coverage on support, is there harm in starting to use C++11 features in desktop-full?

Reply:
Sorry, I'm not really familiar with the history.  Is the above language saying that C++11 is targeted not enough?  Is there an exception for desktop-full?

Reply:
This is saying that all packages should be able to compile with the C++11 flag.

There is a [C++ section below](https://github.com/ros-infrastructure/rep/pull/139/files#diff-41ae062315c36c45afaa5c192b2ca7c5R378) that covers use of modern C++ in APIs. I think that's what @wjwwood is referring to.
There are a couple related threads on discourse [[1]](https://discourse.ros.org/t/c-11-std-shared-ptr-in-public-api/1990) [[2]](https://discourse.ros.org/t/std-shared-ptr-in-the-geometric-shapes-public-api/395)

Reply:
Thanks for the pointers.  Now that C++11 (or newer) is going to be the default standard of the compiler for all of the supported platforms, my feeling is that it is safe to allow C++11 features as part of the desktop-full API.  That's not to say we will change old APIs, just that new ones are free to use it.  Hopefully others from the community can chime in here and express their concerns or support for this change.

Reply:
> Now that C++11 (or newer) is going to be the default standard of the compiler for all of the supported platforms, my feeling is that it is safe to allow C++11 features as part of the desktop-full API. That's not to say we will change old APIs, just that new ones are free to use it.

That is already exactly the case for Kinetic and Lunar afaik.

Reply:
@dirk-thomas I don't think that's true, I believe we'd need to change this from the motivation section:

> Support for C++11 is now a compiler requirement, but the API of the packages included in desktop-full will not use any C++11-specific feature. External packages are encouraged to follow this guideline.

Reply:
That might be written down there but that is not reflecting reality, e.g. the following two examples require C++11 as of Kinetic already:

* https://github.com/ros/roscpp_core/blob/ed7d28e2644e213aca89538c42a56e6ef0af8c39/rostime/CMakeLists.txt#L41
* https://github.com/ros/ros_comm/blob/247459207e20c1da109fc306e58b84d15c4107bd/clients/roscpp/CMakeLists.txt#L5

Reply:
Well that's a difference issue... @mikaelarguedas was holding off on something like https://github.com/ros/pluginlib/pull/67 (which adds new API's to replace boost usage and deprecates boost functions) because of this recommendation.

Reply:
Maybe the difference is, using something that requires C++11 in a header that you export? Both cases you linked don't strictly mean that, one is a test and the other may only be necessary for the cpp files. Whereas, my pr above would require all packages using pluginlib to either have a compiler that defaults to c++11 or turn it on explicitly.

Reply:
Exposing C++11 in a (transitive) header could easily break existing package since the package declaring that header will likely not export the compiler definition and the existing packages will likely not even consider `<name>_DEFINITIONS`.

From this discussion it seems clear that this specific part of the REP needs a much more detailed description / explanation than as it is atm.

Reply:
> since the package declaring that header will likely not export the compiler definition

Why not? This problem could be solved by:

- having all compilers default to C++11 (don't think this is the case even for the platforms melodic will support)
- have a single core package set the C++ standard to 11 in its extras file
  - downstream packages could override, but it would emulate a situate where all compilers default to c++11
- have every package that uses C++11 in a header set the standard to C++11 in its extras file
- have every package that uses C++11 in a header add `-std=c++11` to the `_DEFINITIONS` variable and fix packages that don't consider it

Personally I like the second one (similar to how we default everyone to shared libraries), but I agree it needs more discussion and input from the community.

> From this discussion it seems clear that this specific part of the REP needs a much more detailed description / explanation than as it is atm.

That's why I asked the question 😄.

Reply:
We had extensive discussions in the past (around MoveIt I think) about the side effects and problems exporting those compiler definitions. E.g. different packages exporting different values. The conclusion on that was that option 3 and 4 will likely be problematic on a large scale.

The second option would be something to consider. I was just trying to point out that exposing C++11 in public headers atm requires downstream packages to handle that (and since they don't do that at the moment break them).

Reply:
Quoting https://gcc.gnu.org/gcc-6/changes.html
>> The default mode for C++ is now -std=gnu++14 instead of -std=gnu++98.

Debian Stretch has 6.3.0, Artful has g++ 7.2.0, all others newer

So adding -std=c++11 will be a "downgrade" from default compilation mode for all supported platforms. I vote for having full "permission" for C++11 or even better 14 in all public apis and dropping whatever boost etc is possible

Reply:
@simonschmeisser Thanks for the link, that is helpful.  I also confirmed empirically that gcc 6+ defaults to having (at least) C++11 features.  Given that all of Ubuntu Artful, Bionic, Debian Stretch, and Fedora 28 are going to have versions of gcc newer than 6, that's a pretty good start.   One potentially complicating factor here is `clang++` on macOS, which, as of High Sierra, prints a warning when using C++11 features without passing `-std=c++11`.  It actually still succeed compilation, but the warning might be annoying for the macOS-porting portions of the community.  If we don't care too much about macOS, then I would suggest we go with option number 1 from @wjwwood's post (just use the compiler default).  If we do care about the warnings on macOS, then I think we'll have to go with something like option 2.  Thoughts?

Reply:
Actually, doing a bit more digging on macOS, it seems like if we have the core packages export a `-Wno-c++11-extensions` flag, then High Sierra won't complain, but will compile c++11.  That suggests to me that we should do 1 (use the compiler defaults), with the additional caveat that the core packages would need to ensure that they export the right flag for downstream users not to complain on macOS.  Any further ideas here?

Reply:
I actually think this might not be sufficient, because IIRC some include statements will not work without the `-std=c++11` flag set, e.g. maybe `#include <memory>` or `#include <functional>`?

Reply:
Yeah, you are right about that, I just double checked.  Without the `-std=c++11` flag, *most* things in clang work, but it doesn't know how to parse functional stuff.  So we'll definitely need to provide the flag for macOS to work.

From this, then, I'd suggest that we follow @wjwwood 's option 2, and provide a `std=c++11` flag down in one of the core packages so basically everything inherits it.  This may be a "downgrade" from the default for certain compilers (the newer gcc ones which have C++14 standard), but packages higher up in the stack can always override it.  If everyone is OK with that idea, then I'll update the motivation section to be a lot clearer on this point, and say that we allow C++11 features to be used basically everywhere, including public APIs.